### ISSUE TITLE
☂️ CSS Parser

### ISSUE THREAD

**Issue by ematipico** [biomejs/biome#268]
### Description

This is an umbrella issue to track the works of the CSS parser.

The lexer of the CSS parser should be ready, so now it's time to start working on the parser itself.

We should already have a [**basic** syntax](https://github.com/biomejs/biome/blob/main/xtask/codegen/css.ungram), although it's possible that some grammar or cases are not covered, so *it's highly** suggested to always make sure we have to nodes to represent a specific grammar.

Resources: 
- The original spec https://drafts.csswg.org/css-syntax/#parser-entry-points, it should give us all the information to parse CSS stylesheets correctly
- Use CSS tree https://csstree.github.io/docs/syntax/, which gives visual tools for better understanding parsing rules. e.g. [parsing a declaration](https://csstree.github.io/docs/syntax/#Type:declaration)

Let's remember that **Biome parser needs to be recoverable**, so we have to consider invalid cases and ensure that the parser can recover itself and resume parsing in a fancy matter.

### How to contribute

Before starting, please wait for us to implement the testing infrastructure. In the meantime, you can take a look at the JS parser and JSON parser to understand how the parsing works, the internal APIs, the recovery system, etc.

In this very issue, add a comment saying you want to work on a specific task(s). We will create the issue for you and eventually assign it to:

> [!NOTE]
> In order to have the issue assigned to you, you must comment on the issue.,

### Prior work

You can the previous Rome's (TS) implementation of the parser: https://github.com/rome/tools/blob/archived-js/internal/css-parser/parser/rules.ts

### Before starting

- [x] Create a new test infrastructure for the CSS parser
- [x] Write instructions to run the CSS test infra

### CSS parser tasks

- [x] parse class selector https://csstree.github.io/docs/syntax/#Type:class-selector
- [x] parse id selector https://csstree.github.io/docs/syntax/#Type:id-selector
- [x] parse attribute selector https://csstree.github.io/docs/syntax/#Type:attribute-selector
- [x] parse pseudo element selector 
- [x] parse pseudo classes selector 
- [x] parse simple block https://drafts.csswg.org/css-syntax/#consume-simple-block
- [x] parse declaration https://csstree.github.io/docs/syntax/#Type:declaration
- [x] parse declaration list https://csstree.github.io/docs/syntax/#Type:declaration-list
- [x] parse function https://drafts.csswg.org/css-syntax/#consume-function
- [x] #1664

#### Parsing `at-rules`

At-rules are very strict, and they have a determined syntax. Let us adhere to that.

- [x] parse charset https://developer.mozilla.org/en-US/docs/Web/CSS/@charset 
- [x] parse color-profile https://developer.mozilla.org/en-US/docs/Web/CSS/@color-profile
- [x] #995
- [x] parse counter-style https://developer.mozilla.org/en-US/docs/Web/CSS/@counter-style
- [x] #1303
- [x] #1030
- [x] #1282
- [x] #1262
- [x] #1408
- [x] #1268
- [x] #1320
- [x] #1049
- [x] parse namespace https://developer.mozilla.org/en-US/docs/Web/CSS/@namespace
- [x] #1310
- [x] #1333
- [x] #1334
- [x] #1478
- [x] #1330
- [x] https://github.com/biomejs/biome/issues/1743

#### Ideas

... more to come

**Comment by kubijo** [comment]
Forgive me if this is a dumb question, but isn't there a chance for you to reuse a lot of what lightningcss does? As far as I know they use some rust crates written by Mozilla for Firefox, so it should be a very solid base...

**Comment by ematipico** [comment]
> Forgive me if this is a dumb question, but isn't there a chance for you to reuse a lot of what lightningcss does? As far as I know they use some rust crates written by Mozilla for Firefox, so it should be a very solid base...

We did consider it. While doing some research we reached out to Devon, and he suggested to **not** use that crate (the Mozilla one) because it wouldn't fit our CST.

I also created a PoC some time ago, and I can confirm that unfortunately that crate has some limitations for us.

**Comment by kubijo** [comment]
Oh… ok, that's a shame. So there is no prior art for you then given your architecture?

**Comment by denbezrukov** [comment]
I'm going to return to this task the next monday🙃

**Comment by suxin2017** [comment]
Hello can I join this task, I have written a similar css parser
Here it is
https://github.com/suxin2017/css-tutorial/blob/main/crates/css_tutorial/src/parser.rs

**Comment by denbezrukov** [comment]
> Hello can I join this task, I have written a similar css parser
> 
> Here it is
> 
> https://github.com/suxin2017/css-tutorial/blob/main/crates/css_tutorial/src/parser.rs

It would be great!
If you'd like you can take block parsing.
I'm currently working on at rules and want to prepare our gramma for them. If you'd like to take at rule it's better to want gramma changes.

**Comment by denbezrukov** [comment]
@suxin2017 
I've updated the ticket description. I'd like to suggest to take `font-face`, `font-palette-values` at rules.

**Comment by suxin2017** [comment]
parse simple block https://drafts.csswg.org/css-syntax/#consume-simple-block
 parse declaration https://csstree.github.io/docs/syntax/#Type:declaration
 parse declaration list https://csstree.github.io/docs/syntax/#Type:declaration-list
 parse function https://drafts.csswg.org/css-syntax/#consume-function
I've done that and I'm fixing the failing tests

**Comment by togami2864** [comment]
@denbezrukov @suxin2017 
Can I work on "font-face"??

**Comment by suxin2017** [comment]
If we use all css module level 3 syntax for parsing,
Wouldn't it be better to put the validation of the rules into the analyze module
Because modern css syntax has started to support nesting
https://caniuse.com/css-nesting
It may be difficult for outdated css syntax to support these new features in the future

**Comment by denbezrukov** [comment]
> If we use all css module level 3 syntax for parsing,
> 
> Wouldn't it be better to put the validation of the rules into the analyze module
> 
> Because modern css syntax has started to support nesting
> 
> https://caniuse.com/css-nesting
> 
> It may be difficult for outdated css syntax to support these new features in the future

Could you please give an example of the problem?

**Comment by suxin2017** [comment]
> > If we use all css module level 3 syntax for parsing,
> > Wouldn't it be better to put the validation of the rules into the analyze module
> > Because modern css syntax has started to support nesting
> > https://caniuse.com/css-nesting
> > It may be difficult for outdated css syntax to support these new features in the future
> 
> Could you please give an example of the problem?
It's a discussion, so there's nothing wrong with it so far, but when I was implementing css syntax, I found that there wasn't a standard place to go, Because our grammar and CSS draft (https://drafts.csswg.org/css-syntax/#parser-diagrams) is not consistent, so is the need to read more information

**Comment by ematipico** [comment]
> If we use all css module level 3 syntax for parsing,
> Wouldn't it be better to put the validation of the rules into the analyze module
> Because modern css syntax has started to support nesting
> https://caniuse.com/css-nesting
> It may be difficult for outdated css syntax to support these new features in the future

That's a good point, and I think it's very difficult to find a line between what goes in the parser and what goes in the analyzer.

The CSS parsing rules (the ones from the spec) are very strict, and if you follow them, you end up having a perfect CSS file, and at the end you will need only a few lint rules.

For example:

```css
. something {
  width: calc( 6 / hey )
}
```

If we follow the spec, this would result in a parsing error. `calc()` has a very strict parsing rule. Do we want that in the parser or the linter?

Overall, we have three alternatives:
- strict parsing rules (follow the spec), most of the work is here
- use the **syntax** rules. These are special, non-configurable, lint rules meant to catch syntax errors
- use the **lint** rules. Configurabile, Classic lint rules


In order to be able to have lint rules, we also need **good** CST (nodes), so we can query the CSS correctly.

So, at the end it's a judgement call to where we want to stop the parsing logic and delegate it to the linter.

**Comment by suxin2017** [comment]
I implemented a very rough draft implementation, all the tests have passed, do you feel that such a loose cst will help us

https://github.com/biomejs/biome/pull/1224

**Comment by unvalley** [comment]
I'd like to work on `@font-palette-values`

**Comment by denbezrukov** [comment]
It could be a good first at rule to implement: https://github.com/biomejs/biome/issues/1333

**Comment by yossydev** [comment]
> It could be a good first at rule to implement: #1333

@denbezrukov 
Hello! I would like to work on that issue, is it possible?

**Comment by denbezrukov** [comment]
I've updated the issue to start parsing properties.
source: https://github.com/parcel-bundler/lightningcss/tree/master/src/properties

**Comment by faultyserver** [comment]
I created #1359 to track implementing the CSS-wide keywords as tokens and making them invalid as custom identifiers.

**Comment by faultyserver** [comment]
I've gotten part way through implementing some infrastructure to enable exact parsing of individual properties, but I'm realizing that it will require some changes to the parser itself to support checkpoint rewinding (similar to how the JS parser already works), since there's ambiguity about whether a parsed value is valid until reaching the end of the value definition (i.e., the `;` at the end of the line). I'll probably try to build out that checkpointing in a separate PR first and then utilize it in this work, and then from there we _should_ be able to start parsing a lot of properties directly!

I'll open an issue for tracking that work either later tonight or tomorrow.

**Comment by denbezrukov** [comment]
https://github.com/biomejs/biome/pull/1472#discussion_r1445181445

**Comment by Gami13** [comment]
Could I start working on implementing the other properties in a matter similar to the border one or are you working on a better solution?

**Comment by ematipico** [comment]
@Gami13 do you mean this issue? https://github.com/biomejs/biome/issues/1454

If so, yes, feel free to comment the issue and I will assign it to you

**Comment by denbezrukov** [comment]
I think that we want to keep property parsing as it is now. We want to have properties at a semantic level.

cc @faultyserver

**Comment by denbezrukov** [comment]
@Gami13 
You can try to implement https://github.com/biomejs/biome/issues/1333 if you want.

**Comment by Gami13** [comment]
@ematipico yeah, exactly but for the other properties like align, background etc. assuming the way its done in the border one is close to final, I could work on implementing it for the rest

**Comment by faultyserver** [comment]
> assuming the way its done in the border one is close to final, I could work on implementing it for the rest

@Gami13 as Denis mentioned, after investigating with #1453 and discussing afterward, we are most likely going to keep the parser generic, just parsing component value lists for each property, and then have the _analyzers_ do the actual semantic parsing to get the real values for each rule. There's just too much context and evaluation needed for this to be feasible in a parser, and it will be much easier with an evaluation engine in the analyzer as well.

So, unfortunately, please ignore how the `border` property is being parsed, it will be removed eventually and returned to being generic. The best place to add new features in the parser would be the remaining @-rules like #1333 that Denis mentioned

**Comment by bjoroen** [comment]
Hello,
I would love to get more involved with this project, is there any of the properties issues that would be good for me to start with? Or is there anything else that needs to be done?

**Comment by denbezrukov** [comment]
@bjoroen
I believe it's better to look at https://github.com/biomejs/biome/issues/1285
However, you could take https://github.com/biomejs/biome/pull/1264#discussion_r1433025688

**Comment by bjoroen** [comment]
@denbezrukov I could take #1455 if it's possible.

**Comment by bjoroen** [comment]
You can assign #1743 to me, and I'll take a look at it tomorrow.

**Comment by ematipico** [comment]
@bjoroen you have to comment on the issue, so we can assign that to you. That's how GitHub works
